### PR TITLE
Issue management proposal

### PR DESCRIPTION
**Pull request by npepinpe**
New issue management proposal, largely Kanban inspired, but goes a bit beyond and touches on other things. In progress.

### CODE REVIEW

File: ZEP000-project-management.md [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+Prioritization is done primarily to ease the task of assigning issues to developers (whether self-assigned or by the project manager).
+
+Issues are prioritized in under four different priorities: critical, high, mid, and low.
+
+Critical issues are fast tracked issues which take precedence over everything, and will interrupt the team's current work, e.g. stop the world issues. For example, a bug in production which has a high chance of causing data loss would be critical and take precedence over any other issue.
+
+High priority issues are those which the team will work on, and will be limited to a small number, and referred to as `ready`.
+
+Mid and low priorities are mostly to help the project manager keep track of which issues should become high priority next. Mid priority issues are those which the project manager believe may become high priority in the near future, and are referred to as `planned`. :ow priority those which they believe should be done but most likely will not become high priority anytime soon, and are referred to as `backlog`.
```
Comment (saig0) [inline ZEP000-project-management.md:62]: Fix typo.

```suggestion
Mid and low priorities are mostly to help the project manager keep track of which issues should become high priority next. Mid priority issues are those which the project manager believe may become high priority in the near future, and are referred to as `planned`. Low priority those which they believe should be done but most likely will not become high priority anytime soon, and are referred to as `backlog`.
```

**Review by saig0 — APPROVED** [review]
Sounds good for me :+1: Nice writing :cake: 

Personally, I think this is a lot of text for a relatively simple change :sweat_smile: The core point is something like "we limit the number of open/active issues for better planning by introducing limits/capacities".

File: ZEP000-project-management.md [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+Once an issue has been mid priority for some time, it is automatically downgraded to low priority. Once an issue has been low priority for some time, it is automatically closed. Together they form a rolling backlog, which should help speed up the task of backfilling the pool of high priority issues.
+
+## Assigning
+
+The project manager will mark high priority issues as ready to be worked on. 
+
+## Visualization
+
+A [Codetree](codetree.io) board is used to visualize the current workload of the team, with the following columns:
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply ZEP000-project-management.md]: My idea for using Codetree is that they support multiple repositories in one board. I don't think you can do that with Github projects, can you? I'm open to any other tool, it's just Camunda already has a subscription with Codetree.

File: ZEP000-project-management.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+High priority issues are those which the team will work on, and will be limited to a small number, and referred to as `ready`.
+
+Mid and low priorities are mostly to help the project manager keep track of which issues should become high priority next. Mid priority issues are those which the project manager believe may become high priority in the near future, and are referred to as `planned`. :ow priority those which they believe should be done but most likely will not become high priority anytime soon, and are referred to as `backlog`.
+
+Once an issue has been mid priority for some time, it is automatically downgraded to low priority. Once an issue has been low priority for some time, it is automatically closed. Together they form a rolling backlog, which should help speed up the task of backfilling the pool of high priority issues.
+
+## Assigning
+
+The project manager will mark high priority issues as ready to be worked on.
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply ZEP000-project-management.md:68]: That's defined in the reference guide, but I can add a blurb about it here :+1:

**Comment by npepinpe** [comment]
> Personally, I think this is a lot of text for a relatively simple change

Yes, I agree :see_no_evil: I erred on the side of caution here since it was clear for me, but I wasn't sure if it would be clear to everyone so maybe I explained a little too much :sweat_smile:

File: ZEP000-project-management.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (saig0) [reply ZEP000-project-management.md]: Ok, fair point. Maybe, you can add the reason for Codetree to the document as well :)

**Review by pihme — CHANGES_REQUESTED** [review]
I've read the guide level explanation, and skimmed over the reference level explanation. (meaning I might have missed some detail which is explained later)

I have several issues with the proposal:
I am missing a concept like e.g. error budget to prioritize bugfixes over feature development; here all issues are treated the same, and I think this is wrong.

Also the proposal contains no tracking of progress. E.g. how many issues can we do in a day? Is the number of high prio iseuss growing or shrinking over time? Where in the process do issues get stuck? Is the balance of bugs vs features healthy?

For me the priority of an issue and whether it is planned/ready or worked on are two different aspects, which are orthogonal to each other. In particular, the notion to work only on "high prio" stuff means that low priority stuff never gets done. I think this is a missed opportunity in terms of the broken windows theory. 

I would rather favor some sort of https://personalexcellence.co/blog/big-rocks/ so a sensible mix of high mid and low priorities.

One other question that is on my mind: What is the point of Game Days, Chaos Tests and so on, if they produce only mid-level stuff, which then never get done? I would like to see some provision for being able to fix a problem immediately - even if it is not the most important problem.

I am strongly opposed to any automatic downgrading of issues. Just because we didn't have time to work on an issue doesn't mean that it is less important. I am not opposed to a human doing the downgrading, just about it being done automatically stuff.

Also, yes we can close a low level issue. But then the medic finds an error in the log, and voila, they will reopen a new ticket for the same low level issue again.

The cap of 3 issues per developer is too low, in my experience. I routinely exceed this limit. Sometimes by choice, but more often due to waiting for someone else and then looking for other topics to work on in the meantime.

I didn't get the limit for mid/low priority issues. Is it per team/per developer? Why is it there is we work on high level issues anyway?

The description still sounds like there is a finite amount of issues per team member. Whereas my read of Kanban is that there are work limits for the team and the team picks from that pool. 

This could also imply that developers self-assign to do reviews, instead of what we practice today.

I don't think we need a new role for this; I think the medic can take on the prioritization.

Also, for feature requests from the community somebody should decided whether we actually want to make this change, or not. I feel that is a step which is currently missing. Or maybe not done consistently.

It would be helpful for have time-box recommendations for the categorization and prioritization step. 

It would also be helpful to talk about backporting (when to backport? Who should do it? Is it a new issue or part of the original issue?

It would also help to talk about how Acknowledged errors in StackDriver shall be tracked/closed when working on an issue,

I am sceptical about the breaking work down in smaller chunks part. Sometimes this works well, when you can see the road ahead clearly. Sometimes you have to play it by ear. 
As a trend I hope that we will tackle more challenging topics where the road ahead is not clear. 

I am critical of the "needs more information" label. I feel that it is often used as a crux to procrastinate a decision. I would rather see us being able to make decisions even with incomplete information and treating additional information as optional.

Overall, I feel it is a step in the wrong direction. 

Let me elaborate this a little bit further:
* I feel it elevates priority over common sense
* I feel it has no build-in mechanisms to ensure or even track product health and developer happiness (it proposes a process to do "best effort" work; which is ok; however, we don't know/track whether the best effort will be enough to make the product healthier; and if best effort is not enough, then developers will become frustrated/disheartened; we can never do more than best effort with the resources we have, but I think we need to gather information whether things largely get better or worse in order to plan resources for the future)
* I feel it doesn't change enough. I want to see a bigger change towards a work pool for the team, vs. work pools for individual developers. I also want to see a clearer separation of bug vs, feature issues.

**Comment by npepinpe** [comment]
All valid criticism, which is definitely helpful, and I'll address it point by point. However, a lot of these simply point out "this is problematic" without really offering any solution. You clearly have more experience than I do in this field, so it would help a lot if you could not only point out issues with the proposal, but also provide suggestions on how to solve these.

For example:

> I am critical of the "needs more information" label. I feel that it is often used as a crux to procrastinate a decision. I would rather see us being able to make decisions even with incomplete information and treating additional information as optional.

This is a valid point complaint, but doesn't offer any actionable solution, only expresses a desired outcome.

So, anyway, let's go point by point over the review then:

> I am missing a concept like e.g. error budget to prioritize bugfixes over feature development; here all issues are treated the same, and I think this is wrong.

In my opinion, the error budget falls outside of how we manage issues - in a way, it supersedes how we manage issues. In fact, the proposal doesn't go too in depth on how issues are prioritized, it simply describes the various columns into which they may be prioritized, and gives a few examples. I'm not sure making an exhaustive list of all the factors which may influence prioritization is useful, and I'm not even sure it's possible to describe an objective way of doing it. I understand your point, though, as it indirectly will impact which issues make it into the pool of ready issues. I can gladly add a mention of it, but I still feel it's a bit outside of scope here.

> Also the proposal contains no tracking of progress. E.g. how many issues can we do in a day? Is the number of high prio iseuss growing or shrinking over time? Where in the process do issues get stuck? Is the balance of bugs vs features healthy?

Very true, and again I thought it was out of scope for now. This sounds like a way to optimize the process, and I would be happy to do this as a second iteration. I'm also not aware of any tools which would allow us to do this right now, so if you know of any please share. I proposed Codetree here as it's a board which supports multi-repo, but I'm not sure if they provide any kind of historical statistics or tracking. Again, happy to try out another tool, but I also think this can be a second iteration.

> For me the priority of an issue and whether it is planned/ready or worked on are two different aspects, which are orthogonal to each other. In particular, the notion to work only on "high prio" stuff means that low priority stuff never gets done. I think this is a missed opportunity in terms of the broken windows theory.

So here I think there was a misunderstanding in the terminology, and it's probably my fault. I should just stop using high/mid/low priorities, and just used backlog/planned/ready. There will be no high/mid/low priorities any more. I used these cause I had hoped it would help people understand how the transition would go, but I guess that didn't work. The idea is that the team works on issues in the ready column - the planned and backlog columns are purely for me to make things manageable when it comes to filling the ready/planned column (i.e. to fill the ready column I look at the planned, to fill the planned at the backlog). Anything in ready should be approached as having the same priority, whether it's what we've been calling a low or high priority issue. So the proposal puts a lot of the responsibility on how well we balance small and large issues on my shoulders. My current hypothesis is that if we keep good velocity and we can breakdown issues properly, then it shouldn't be too difficult to achieve a good balance, where we can both reduce our tech debt, fix bugs, and also achieve our OKRs and larger topics.

> I would rather favor some sort of https://personalexcellence.co/blog/big-rocks/ so a sensible mix of high mid and low priorities.

I am not aware of this, will take a look at it today :+1: 

> One other question that is on my mind: What is the point of Game Days, Chaos Tests and so on, if they produce only mid-level stuff, which then never get done? I would like to see some provision for being able to fix a problem immediately - even if it is not the most important problem.

Genuine question, why do you want to fix a problem immediately if it isn't the most important problem? If I have X very important problem, and the game day discovers 1 not so important, why do you want a provision to immediately fix that one? The primary goal of game day is to prepare us to resolve incidents, not to fix issues. Chaos tests are a bit different, but these should simply become part of our testing strategy, and issues discovered there will get prioritized along with other bug fixes. But maybe I misunderstood your point, it would help to elaborate why you would want to see this, pros and cons, etc.

> I am strongly opposed to any automatic downgrading of issues. Just because we didn't have time to work on an issue doesn't mean that it is less important. I am not opposed to a human doing the downgrading, just about it being done automatically stuff.

I disagree. To me it sounds like we thought something was important at the time, but as more things piled on, or new things were discovered, the stale issues are simply not as important as before. It's simply a matter of perspective - new issues come to light, and in a perfect world, with very few issues, would cause us to reprioritize. This is at the moment simply not feasible, there would be too much overhead there. I will concede that there's a chance things fall through the crack, and one thing I thought for this would be to have a weekly blocked session where I would look at new stale issues to see if they should be unmarked, so they don't get downgraded.

> Also, yes we can close a low level issue. But then the medic finds an error in the log, and voila, they will reopen a new ticket for the same low level issue again.

I don't see the issue here. Ideally they'd reopen the old issue, but it's fine if they create a new one. Maybe you can elaborate on why this is a problem.

> The cap of 3 issues per developer is too low, in my experience. I routinely exceed this limit. Sometimes by choice, but more often due to waiting for someone else and then looking for other topics to work on in the meantime.

This is one of the open questions at the bottom (maybe not the best place), where I admit all capacities mentioned are provisional and I don't know what would be good limits, or how to find them - suggestions are highly appreciated. Saying it's too low for you is helpful for me, but even better would be if you could provide a suggestion on a team-wide capacity, and the reasoning behind the suggestion.

> I didn't get the limit for mid/low priority issues. Is it per team/per developer? Why is it there is we work on high level issues anyway?

Again, misunderstanding on the terminology. There is no more high/mid/low, that's my fault for writing this. The planned and backlog columns are there for me, so I can keep a handle on how to replenish the ready column (and the planned in the case of backlog). They're really just to keep it manageable for me, and can be ignored by the team (though of course it might still be useful to peek at it).

> The description still sounds like there is a finite amount of issues per team member. Whereas my read of Kanban is that there are work limits for the team and the team picks from that pool.

You're correct, but in the past the team mentioned they wanted individual capacities - you can extrapolate the Kanban capacity from the individual one, of course. This is also an open question at the bottom - I'm happy to have a team wide capacity per column as well if the team's on board, but would be concerned that people may self assign too many (though, I suppose if we're self assigning, this shouldn't happen).

> This could also imply that developers self-assign to do reviews, instead of what we practice today.

I'm not sure I follow - do you mean to suggest we should self-assign for reviews? If so, can you elaborate on why this is a good idea? I'm still strongly in favor of picking your reviewer before you start so you can provide them with context and they are part of the design phase, which should reduce frictions during the review phase.

> I don't think we need a new role for this; I think the medic can take on the prioritization.

Do you mean the categorization? The project manager (me, but hopefully one day a PM) should do the prioritization. If you meant prioritization, can you elaborate why the medic should do it? Or would it be scoped only to bugs?

> Also, for feature requests from the community somebody should decided whether we actually want to make this change, or not. I feel that is a step which is currently missing. Or maybe not done consistently.

Right now, this is deferred to the stakeholders/product manager. This is a long standing issue, but it's not one I'm willing to solve by simply pushing the responsibility onto me. I want it to be painfully clear to our product team that we need a PM, so me taking more responsibility there would probably be counter productive.

> It would be helpful for have time-box recommendations for the categorization and prioritization step.

True, and I'm happy to hear suggestions - I imagine any recommendation I make would get torn apart :wink:, but I can gladly give it a try.

> It would also be helpful to talk about backporting (when to backport? Who should do it? Is it a new issue or part of the original issue?

Hm, to me this seemed out of scope, as the proposal is more about which issues we work on and when, not really about how we work on issues. Can you elaborate why you feel this falls within this scope? Or is it maybe more of an issue you'd like to see tackled, but are unsure of where it should be done?

> It would also help to talk about how Acknowledged errors in StackDriver shall be tracked/closed when working on an issue.

Agreed, right now we simply don't track it because of the overwhelming number of issues that were opened. Do you see an advantage to marking issues as solved in Stackdriver? I find Stackdriver to be useful to discover errors, but I don't see much advantage in tracking the status there. One thing for sure is these issues should be tackled much sooner than we do now, as we only keep logs for a month. Maybe I can work something in there regarding those...

> I am sceptical about the breaking work down in smaller chunks part. Sometimes this works well, when you can see the road ahead clearly. Sometimes you have to play it by ear.

Agreed, but we should always try to break it down as much as possible. This is key to me to achieve a good balance between refactoring/reducing tech debt, bug fixing, and features. 

> As a trend I hope that we will tackle more challenging topics where the road ahead is not clear.

How would you plan such things, while also keeping a good balance between bug fixing and smaller refactoring issues? Maybe I misunderstand your statement? I mean, we will always have challenging topics where we don't know, and these will always require an initial issue where we need to figure out what can be done - once that's done, then we can do a break down. I think I touched on this in the reference guide (but maybe I forgot).

> I am critical of the "needs more information" label. I feel that it is often used as a crux to procrastinate a decision. I would rather see us being able to make decisions even with incomplete information and treating additional information as optional.

I understand, and I'm sure it happens some times, but I think it's unavoidable. There will be users who simply do not provide us enough information to do much, or even to understand what the problem is. How can you make an informed decision if you don't understand what the issue is about? What kind of decision could be made? What is the advantage of making a decision with incomplete information? I understand sometimes we will simply have to go ahead and do it, and I think I touched on that also in the reference level guide, but I strongly believe we need this label as I know for a fact we will have issues where there is no advantage in making a decision with incomplete information. I imagine here you refer to the current state, where I sometimes don't know what decision to make as I feel I have not enough information, and this is perceived as indecisiveness (and probably is, to be honest). I hope there it's just a matter of me getting more experience, and again I'm happy to hear any suggestions, or constructive feedback.

>    I feel it elevates priority over common sense

It's unclear what you mean by this here, maybe you can elaborate, and even better would be to suggest an alternative.

>    I feel it has no build-in mechanisms to ensure or even track product health and developer happiness (it proposes a process to do "best effort" work; which is ok; however, we don't know/track whether the best effort will be enough to make the product healthier; and if best effort is not enough, then developers will become frustrated/disheartened; we can never do more than best effort with the resources we have, but I think we need to gather information whether things largely get better or worse in order to plan resources for the future)

I'd be happy to hear any proposal there :slightly_smiling_face: 

>    I feel it doesn't change enough. I want to see a bigger change towards a work pool for the team, vs. work pools for individual developers. I also want to see a clearer separation of bug vs, feature issues.

We haven't really worked on features in almost a year, though, and most likely won't in the near future, as we're completely over our error budget. But maybe you can elaborate here - what would you like to see? Examples would be very helpful.

Overall, it sounds to me like you have very strong ideas about what you don't want/want, but weren't very forthcoming with solutions. Would you like to work on your own proposal? I'm happy to delegate this, as all this management stuff is very new to me, and I can understand me learning on the job may be frustrating for all of you who end up being my test subjects, in a way. We can look into allocating time for you to work on your own proposal.

**Comment by pihme** [comment]
@npepinpe 
You are right to point out that my feedback doesn't contain many suggestions. I would like to hear the other team members first to see where we find common ground.

I am happy to make suggestions once we have a clearer picture of where we are headed.

But reading your response, I am not sure I fully understood your proposal in the first place. So I guess I better listen more and talk less.

Anyways, some suggestions up front:
* Try it out first, then collect feedback, then refine, then write up a ZEP
* we are now in a place where we sometimes have time at the end of the triage meeting. We can use this time to follow up on the "needs information" issues. Let's say we didn't get any information after two weeks. Then we can look at the ticket again. Either there is nothing we can do, then we should close it. Or maybe we want to look at it closer. Then we should assign a best guess category. In any case, I would prefer this to be a manual task, not an automated mechanism
* Talk to stakeholder which information they want and at what level of details. In particular, I do not know what the expectation is on working on OKRs vs. reducing the backlog. Maybe they expect us to take case of the backlog and work on OKR only in the remaining free time, or they expect us to elevate work on the backlog to the OKR sheet. I really don't know what the modus operandi is there
* At the begin of the quarter, go through the "Needs stakeholder input" ticket and find those that are relevant for the quarter. Then get stakeholder input for those.
* At the begin of the month, get the team together and we discuss which issues we want to tackle for the next alpha release. This will hopefully give you more engagement and buy in from the team
* At the end of the month reset all ready issues of the last month and start anew
* time box for categorization: 15 minutes on average (meaning there can be issues that take longer while others are shorter)

Also, some clarifications on where I think we have misunderstood each other:

* Why I want to fix a problem immediately (e.g. after something was discovered in a GameDay)?

There are several aspects to this. For me the goal of the Game day is - as you said - to prepare us to resolve incidents. Part of the outcome of a game day is that we discover where there is friction in the process. Once we identified the friction, we should remove it. First, having less friction will help us in the next Game Day and in an actual incident. Also this makes developers happier, because they found a problem and fixed it. Maybe you are thinking of bugs here, whereas I am thinking of missing documentation, unclear log messages and so on. These will probably not be the most important thing. Currently, I feel (and this is more  a feeling; I have not validated this) many of our insights become issues which then appear on the backlog. But if it takes a year to improve things, what is the point of rehearsing incident responses? And with your suggestions the worst case is that the improvement will be automatically downgraded and subsequently closed. 

More in general, I have seen very often how the obsession with "the most important thing to work on" resulted in a huge mess, because the little things accumulated and made everything messy. To take a metaphor from real life, whatever your priorities in life are, brushing your teeth is probably not your top priorities. And yes, you might skip it once in a while, but you still do it. Because this is part of hygiene and if hygiene is neglected, it has bad consequences. 

What I have experienced in the past is that people were so focused on the most important problems that they neglected hygiene. And over time lack of hygiene became the most important problem.

* Why I am against auto-closing issues ?
I am not dead against it. I am opposed to it especially in connection with not tracking project health and trends. I would mind it less, if a human closes it, because then it is a deliberate decision. But just closing it automatically after time has run out is just too random for my liking. Also, if you expect the medic to reopen closed issues, then the medic still has to look at the backlog + all the closed issues. So nothing changes for the medic.

* Why I think tracking closed bugs in Stackdriver is important?
This is the only way I see to verify that the errors are indeed fixed. If we mark it as closed, Stackdriver will alert us again if it occurs again (at least I hope this is what will happen). If we don't mark it as closed then Stackdriver will think we are still working on it and will not alert us again even if the error occurs again in the version, where we thought we had fixed it.

* When I talk about bug vs. feature issues I include anything that is not a bugfix in the feature issue bucket. So in this sense e.g. the testbench is a new feature. I guess my line of thinking is: do we fix existing code or write new code. Because all new code will also produce new bugs and more work to maintain it. So often the best way to reduce the flood of incoming issues is to fix the bugs you already know about. 

* Regarding self assigning for reviews: This is mostly to follow the pull principle of Kanban. If developers get assigned to reviews they have no control over when this happens and how many they have to do. There have also been complaints in the past that some people are more involved with reviews than others. This can clog up reviews needlessly. The benefit of self assigning for reviews is that then you increase the flow of issues (assuming a team Kanban board, not a personal one)

* The backporting thing would be interesting if this is considered a column in the normal flow of working on an issue, or whether the backports are separate issues.

* And finally and maybe most importantly, about the things you declared out of scope (bugs vs. features, OKRs vs, non OKR, tracking project health). I understand that you want to declare them out of scope. Personally, for me these are the more interesting topics to talk about. The mechanics of how we deal with issues are not so interesting for me without looking at the broader picture: What is the goal of managing issues? Do we want to move fast and break things? Do we ever want to have zero bugs? Do we want to use the backlog as a store of information in order not to forget things? Whatever the goals are, are we on the right track? Do we achieve them?

Without this context I have a hard time evaluating the pros and cons of the proposal. And I have a suspicion that once we are clearer on why we are managing issues, this will effect how we manage issues.

**Comment by npepinpe** [comment]
@pihme Sorry I haven't had time to work on this again this week - let's block some time next week to work on it together.

**Comment by npepinpe** [comment]
As we discussed (sorry it took so long to get back), I'm going to close this for now - we'll rather approach it by doing small iterations and trying things out, and we can formalize it once we've found something we're happy with.

**Pull request closed without merging**